core/ordersync_subprotocols.go

```diff
@@ -0,0 +1,174 @@
+		if !ok {
+			return nil, fmt.Errorf("FilteredPaginationSubProtocol received request with wrong metadata type (got %T)", req.Metadata)
+		}
+	}
+	// It's possible that none of the orders in the current page match the filter.
+	// We don't want to respond with zero orders, so keep iterating until we find
+	// at least some orders that match the filter.
+	filteredOrders := []*zeroex.SignedOrder{}
+	var snapshotID string
+	var currentPage int
+	for currentPage = metadata.Page; len(filteredOrders) == 0; currentPage += 1 {
+		// Get the orders for this page.
+		ordersResp, err := p.app.GetOrders(currentPage, p.perPage, metadata.SnapshotID)
+		if err != nil {
+			return nil, err
+		}
+		snapshotID = ordersResp.SnapshotID
+		if len(ordersResp.OrdersInfos) == 0 {
+			// No more orders left.
+			break
+		}
+		// Filter the orders for this page. If none of them match the filter, we continue
+		// on to the next page.
+		for _, orderInfo := range ordersResp.OrdersInfos {
+			if matches, err := p.orderFilter.MatchOrder(orderInfo.SignedOrder); err != nil {
+				return nil, err
+			} else if matches {
+				filteredOrders = append(filteredOrders, orderInfo.SignedOrder)
+			}
+		}
+	}
+
+	return &ordersync.Response{
+		Orders:   filteredOrders,
+		Complete: len(filteredOrders) == 0,
+		Metadata: &FilteredPaginationResponseMetadata{
+			Page:       currentPage,
+			SnapshotID: snapshotID,
+		},
+	}, nil
+}
+
+// HandleOrderSyncResponse handles the orders for one page by validating them, storing them
+// in the database, and firing the appropriate events. It also returns the next request to
+// be sent. This is the implementation for the "requester" side of the subprotocol.
+func (p *FilteredPaginationSubProtocol) HandleOrderSyncResponse(ctx context.Context, res *ordersync.Response) (*ordersync.Request, error) {
+	if res.Metadata == nil {
+		return nil, errors.New("FilteredPaginationSubProtocol received response with nil metadata")
+	}
+	metadata, ok := res.Metadata.(*FilteredPaginationResponseMetadata)
+	if !ok {
+		return nil, fmt.Errorf("FilteredPaginationSubProtocol received response with wrong metadata type (got %T)", res.Metadata)
+	}
+	filteredOrders := []*zeroex.SignedOrder{}
+	for _, order := range res.Orders {
+		if matches, err := p.orderFilter.MatchOrder(order); err != nil {
+			return nil, err
+		} else if matches {
+			filteredOrders = append(filteredOrders, order)
+		} else if !matches {
+			p.app.handlePeerScoreEvent(res.ProviderID, psReceivedOrderDoesNotMatchFilter)
+		}
+	}
+	validationResults, err := p.app.orderWatcher.ValidateAndStoreValidOrders(ctx, res.Orders, false, p.app.chainID)
+	if err != nil {
+		return nil, err
+	}
+	for _, acceptedOrderInfo := range validationResults.Accepted {
+		if acceptedOrderInfo.IsNew {
+			log.WithFields(map[string]interface{}{
+				"orderHash": acceptedOrderInfo.OrderHash.Hex(),
+				"from":      res.ProviderID.Pretty(),
+			}).Info("received new valid order from peer via ordersync")
```

Comment:
```suggestion
			}).Info("received new valid order from peer")
```

Let's not modify the log message since the data pipeline is querying for it exactly. If we want to distinguish between those discovered via `ordersync` and `gossipsub`, let's add a `source` field to the log.